src/Microsoft.SqlTools.ServiceLayer/Agent/Notebooks/Exec-Notebook.ps1

```diff
@@ -0,0 +1,211 @@
+param(
+    [System.String]
+    $TemplateNotebook = $(throw New-Object System.Exception "TemplateNotebook cannot be null"),
+
+    [System.String]
+    $DatabaseInstance,
+
+    [System.String]
+    $DatabaseServer,
+
+    [System.String]
+    $DatabaseUsername,
+
+    [System.String]
+    $DatabasePassword
+)
+<#
+    Parses a DataTable returned by Invoke-SQL Command into a Jupyter Notebook output cell.
+#>
+function ParseTableToNotebookOutput {
+    param (
+        [System.Data.DataTable]
+        $DataTable,
+
+        [int]
+        $CellExecutionCount
+    )
+    $TableHTMLText = "<table>"
+    $TableSchemaFeilds = @()
+    $TableHTMLText += "<tr>"
+    foreach ($ColumnName in $DataTable.Columns) {
+        $TableSchemaFeilds += @(@{name = $ColumnName.toString() })
+        $TableHTMLText += "<th>" + $ColumnName.toString() + "</th>"
+    }
+    $TableHTMLText += "</tr>"
+    $TableSchema = @{ }
+    $TableSchema["fields"] = $TableSchemaFeilds
+
+    $TableDataRows = @()
+    foreach ($Row in $DataTable) {
+        $TableDataRow = [ordered]@{ }
+        $TableHTMLText += "<tr>"
+        $i = 0
+        foreach ($Cell in $Row.ItemArray) {
+            $TableDataRow[$i.ToString()] = $Cell.toString()
+            $TableHTMLText += "<td>" + $Cell.toString() + "</td>"
+            $i++
+        }
+        $TableHTMLText += "</tr>"
+        $TableDataRows += $TableDataRow
+    }
+
+    $TableDataResource = @{ }
+    $TableDataResource["schema"] = $TableSchema
+    $TableDataResource["data"] = $TableDataRows
+    $TableData = @{ }
+    $TableData["application/vnd.dataresource+json"] = $TableDataResource
+    $TableData["text/html"] = $TableHTMLText
+    $TableOutput = @{ }
+    $TableOutput["output_type"] = "execute_result"
+    $TableOutput["data"] = $TableData
+    $TableOutput["metadata"] = @{ }
+    $TableOutput["execution_count"] = $CellExecutionCount
+    return $TableOutput
+}
+
+
+function ParseQueryErrorToNotebookOutput {
+    param (
+        $QueryError
+    )
+    <#
+    Following the current syntax of errors in T-SQL notebooks from ADS
+    #>
+    $ErrorString = "Msg " + $QueryError.Exception.InnerException.Number +
+    ", Level " + $QueryError.Exception.InnerException.Class +
+    ", State " + $QueryError.Exception.InnerException.State +
+    ", Line " + $QueryError.Exception.InnerException.LineNumber +
+    "`r`n" + $QueryError.Exception.Message
+    
+    $ErrorOutput = @{ }
+    $ErrorOutput["output_type"] = "error"
```

Comment:
Is this how we send errors from the notebook or does it come out as `'application/vnd.jupyter.stderr'`?

Reply:
Actually - checked the SQLKernel code and you're matching it. Please ignore.